### ISSUE TITLE
Incorrect output dtype for operations on arrays and scalars

### ISSUE THREAD

**Issue by ehsantn** [numba/numba#4449]
## Reporting a bug

- [x] I am using the latest ~~released~~ master version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Looks like output dtype is incorrect when an operation has array and scalar inputs (see below). This happens for both integer and float types, and seems different than the [integer typing semantic issue](https://numba.pydata.org/numba-doc/dev/proposals/integer-typing.html).

np.testing functions don't seem to check for dtype matching, so this is easy to miss. I came across this issue since pd.testing functions do check for dtype matching.
```python
In [36]: A = np.arange(5, dtype=np.int32)

In [37]: numba.njit(lambda A: A+1)(A).dtype
Out[37]: dtype('int64')

In [38]: (A+1).dtype
Out[38]: dtype('int32')
```
<details><summary>numba -s output</summary>
<p>
__Time Stamp__
2019-08-14 15:46:31.280166

__Hardware Information__
Machine                                       : x86_64
CPU Name                                      : skylake
CPU count                                     : 16
CPU Features                                  :
64bit adx aes avx avx2 bmi bmi2 clflushopt cmov cx16 f16c fma fsgsbase invpcid
lzcnt mmx movbe pclmul popcnt prfchw rdrnd rdseed rtm sahf sgx sse sse2 sse3
sse4.1 sse4.2 ssse3 xsave xsavec xsaveopt xsaves

__OS Information__
Platform                                      : Darwin-18.7.0-x86_64-i386-64bit
Release                                       : 18.7.0
System Name                                   : Darwin
Version                                       : Darwin Kernel Version 18.7.0: Thu Jun 20 18:42:21 PDT 2019; root:xnu-4903.270.47~4/RELEASE_X86_64
OS specific info                              : 10.14.6   x86_64

__Python Information__
Python Compiler                               : Clang 4.0.1 (tags/RELEASE_401/final)
Python Implementation                         : CPython
Python Version                                : 3.7.3
Python Locale                                 : en_US UTF-8

__LLVM information__
LLVM version                                  : 8.0.0

__CUDA Information__
CUDA driver library cannot be found or no CUDA enabled devices are present.
Error class: <class 'numba.cuda.cudadrv.error.CudaSupportError'>

__ROC Information__
ROC available                                 : False
Error initialising ROC due to                 : No ROC toolchains found.
Available Toolchains                          : ROC command line tools
No HSA Agents found, encountered exception when searching:
Error at driver init:

HSA is not currently supported on this platform (darwin).
:

__SVML Information__
SVML state, config.USING_SVML                 : False
SVML library found and loaded                 : False
llvmlite using SVML patched LLVM              : True
SVML operational                              : False

__Threading Layer Information__
TBB Threading layer available                 : False
+--> Disabled due to                          : Unknown import problem.
OpenMP Threading layer available              : False
+--> Disabled due to                          : Unknown import problem.
Workqueue Threading layer available           : True

__Numba Environment Variable Information__
NUMBA_DEBUG_ARRAY_OPT                         : 1

__Conda Information__
conda_build_version                           : 3.18.8
conda_env_version                             : 4.7.11
platform                                      : osx-64
python_version                                : 3.7.3.final.0
root_writable                                 : True

__Current Conda Env__
_tflow_select             2.3.0                       mkl
absl-py                   0.7.1                    py37_0
appnope                   0.1.0                    py37_0
arrow-cpp                 0.11.1           py37hcacac7f_1
astor                     0.8.0                    py37_0
astroid                   2.2.5                    py37_0
atomicwrites              1.3.0                    py37_1
attrs                     19.1.0                   py37_1
backcall                  0.1.0                    py37_0
blas                      1.0                         mkl
bleach                    3.1.0                    py37_0
bodo                      0+untagged.2706.g0d319c5.dirty           dev_0    <develop>
boost                     1.67.0                   py37_4
bzip2                     1.0.8                h1de35cc_0
c-ares                    1.15.0            h1de35cc_1001
ca-certificates           2019.5.15                     1
cctools                   895                           1
certifi                   2019.6.16                py37_1
clang                     4.0.1                         1
clang_osx-64              4.0.1               h1ce6c1d_16
clangxx                   4.0.1                         1
clangxx_osx-64            4.0.1               h22b1bf0_16
cmake                     3.14.0               haff7e42_0
compiler-rt               4.0.1                hcfea43d_1
dbus                      1.13.6               h90a0687_0
decorator                 4.4.0                    py37_1
defusedxml                0.6.0                      py_0
entrypoints               0.3                      py37_0
expat                     2.2.6                h0a44026_0
gast                      0.2.2                    py37_0
gettext                   0.19.8.1             h15daf44_3
gflags                    2.2.2                h0a44026_0
gfortran_osx-64           4.8.5                h22b1bf0_8
glib                      2.56.2               hd9629dc_0
glog                      0.3.5                h0a44026_1
grpcio                    1.16.1           py37h044775b_1
h5py                      2.9.0            py37h3134771_0
hdf5                      1.10.4               hfa1e0ec_0
icu                       58.2                 h4b95b61_1
importlib_metadata        0.17                     py37_1
intel-openmp              2019.4                      233
ipykernel                 5.1.1            py37h39e3cac_0
ipyparallel               6.2.4                    py37_0
ipython                   7.6.1            py37h39e3cac_0
ipython_genutils          0.2.0                    py37_0
ipywidgets                7.5.0                      py_0
isort                     4.3.21                   py37_0
jedi                      0.13.3                   py37_0
jinja2                    2.10.1                   py37_0
joblib                    0.13.2                   py37_0
jpeg                      9b                   he5867d9_2
jsonschema                3.0.1                    py37_0
jupyter                   1.0.0                    py37_7
jupyter_client            5.3.1                      py_0
jupyter_console           6.0.0                    py37_0
jupyter_core              4.5.0                      py_0
keras-applications        1.0.8                      py_0
keras-preprocessing       1.1.0                      py_1
krb5                      1.16.1               hddcf347_7
lazy-object-proxy         1.4.1            py37h1de35cc_0
ld64                      274.2                         1
libboost                  1.67.0               hebc422b_4
libcurl                   7.65.2               h051b688_0
libcxx                    4.0.1                hcfea43d_1
libcxxabi                 4.0.1                hcfea43d_1
libedit                   3.1.20181209         hb402a30_0
libevent                  2.1.8                ha12b0ac_0
libffi                    3.2.1                h475c297_4
libgfortran               3.0.1                h93005f0_2
libiconv                  1.15                 hdd342a3_7
libpng                    1.6.37               ha441bb4_0
libprotobuf               3.8.0                hd9629dc_0
libsodium                 1.0.16               h3efe00b_0
libssh2                   1.8.2                ha12b0ac_0
llvm                      4.0.1                         1
llvm-lto-tapi             4.0.1                         1
llvm-openmp               4.0.1                hcfea43d_1
llvmlite                  0.29.0dev0              py37_36    numba/label/dev
lz4-c                     1.8.1.2              h1de35cc_0
markdown                  3.1.1                    py37_0
markupsafe                1.1.1            py37h1de35cc_0
maven                     3.6.0                         0
mccabe                    0.6.1                    py37_1
mistune                   0.8.4            py37h1de35cc_0
mkl                       2019.4                      233
mkl-service               2.0.2            py37h1de35cc_0
mkl_fft                   1.0.12           py37h5e564d8_0
mkl_random                1.0.2            py37h27c97d8_0
mock                      3.0.5                    py37_0
more-itertools            7.0.0                    py37_0
mpi                       1.0                       mpich
mpi4py                    3.0.1            py37h77202c6_0    conda-forge
mpich                     3.2.1                h12d9a10_8
nbconvert                 5.5.0                      py_0
nbformat                  4.4.0                    py37_0
ncurses                   6.1                  h0a44026_1
notebook                  6.0.0                    py37_0
numba                     0.46.0.dev0+12.g37ab1420f           dev_0    <develop>
numpy                     1.16.4           py37hacdab7b_0
numpy-base                1.16.4           py37h6575580_0
openjdk                   8.0.152              h393ad39_1
openssl                   1.1.1c               h1de35cc_1
packaging                 19.0                     py37_0
pandas                    0.25.0           py37h0a44026_0
pandoc                    2.2.3.2                       0
pandocfilters             1.4.2                    py37_1
parso                     0.5.0                      py_0
pcre                      8.43                 h0a44026_0
pexpect                   4.7.0                    py37_0
pickleshare               0.7.5                    py37_0
pip                       19.1.1                   py37_0
pluggy                    0.12.0                     py_0
prometheus_client         0.7.1                      py_0
prompt_toolkit            2.0.9                    py37_0
protobuf                  3.8.0            py37h0a44026_0
ptyprocess                0.6.0                    py37_0
py                        1.8.0                    py37_0
py-boost                  1.67.0           py37h6440ff4_4
py4j                      0.10.7                   py37_0
pyarrow                   0.11.1           py37h0a44026_0
pygments                  2.4.2                      py_0
pylint                    2.3.1                    py37_0
pyparsing                 2.4.0                      py_0
pyqt                      5.9.2            py37h655552a_2
pyrsistent                0.14.11          py37h1de35cc_0
pyspark                   2.4.3                      py_0
pytest                    5.0.1                    py37_0
python                    3.7.3                h359304d_0
python-dateutil           2.8.0                    py37_0
pytz                      2019.1                     py_0
pyzmq                     18.0.0           py37h0a44026_0
qt                        5.9.7                h468cd18_1
qtconsole                 4.5.2                      py_0
readline                  7.0                  h1de35cc_5
rhash                     1.3.8                ha12b0ac_0
scala                     2.11.1                        0    anaconda-cluster
scikit-learn              0.21.2           py37h27c97d8_0
scipy                     1.3.0            py37h1410ff5_0
send2trash                1.5.0                    py37_0
setuptools                41.0.1                   py37_0
sip                       4.19.8           py37h0a44026_0
six                       1.12.0                   py37_0
snappy                    1.1.7                he62c110_3
sqlite                    3.29.0               ha441bb4_0
tensorboard               1.13.1           py37haf313ee_0
tensorflow                1.13.1          mkl_py37h70c3834_0
tensorflow-base           1.13.1          mkl_py37h66b1bf0_0
tensorflow-estimator      1.13.0                     py_0
termcolor                 1.1.0                    py37_1
terminado                 0.8.2                    py37_0
testpath                  0.4.2                    py37_0
thrift-cpp                0.11.0               hd79cdb6_3
tk                        8.6.8                ha441bb4_0
tornado                   6.0.3            py37h1de35cc_0
traitlets                 4.3.2                    py37_0
wcwidth                   0.1.7                    py37_0
webencodings              0.5.1                    py37_1
werkzeug                  0.15.4                     py_0
wheel                     0.33.4                   py37_0
widgetsnbextension        3.5.0                    py37_0
wrapt                     1.11.2           py37h1de35cc_0
xz                        5.2.4                h1de35cc_4
zeromq                    4.3.1                h0a44026_3
zipp                      0.5.1                      py_0
zlib                      1.2.11               h1de35cc_3
zstd                      1.3.7                h5bba6e5_0
--------------------------------------------------------------------------------

</p>
</details>

**Comment by stuartarchibald** [comment]
Thanks for the report, IIRC this is known behaviour and it is good to have it recorded in an issue. Whilst not exactly the same, CUDA suffers particularly from literals being interpreted as widest type for a numerical domain.

**Comment by ehsantn** [comment]
I think this case is different than literal and scalar cases, since array's dtype should always have higher priority than scalar dtypes.
